metals/src/main/scala/scala/meta/internal/parsing/ClassFinder.scala

```diff
@@ -4,32 +4,74 @@ import scala.annotation.tailrec
 
 import scala.meta.Defn
 import scala.meta.Pkg
+import scala.meta.Position
 import scala.meta.Tree
-import scala.meta.inputs.Position
 import scala.meta.internal.metals.MetalsEnrichments._
 import scala.meta.io.AbsolutePath
 
 import org.eclipse.{lsp4j => l}
 
+final case class ClassWithPos(
+    name: String,
+    pos: Position
+)
+
 class ClassFinder(trees: Trees) {
 
   def findClass(path: AbsolutePath, pos: l.Position): Option[String] =
-    findClass(path, pos, true)
+    findClass(path, pos, checkInnerClasses = true).filter(_.nonEmpty)
 
   def findTasty(path: AbsolutePath, pos: l.Position): Option[String] =
-    findClass(path, pos, false).filter(_.nonEmpty).map(_.stripSuffix("$"))
+    findClass(path, pos, checkInnerClasses = false)
+      .filter(_.nonEmpty)
+      .map(_.stripSuffix("$"))
+
+  def findAllClasses(
```

Comment:
Maybe:
```scala

  def findAllClasses(
      path: AbsolutePath,
      checkInnerClasses: Boolean
  ): Option[List[ClassWithPos]] = {
    for {
      tree <- trees.get(path)
    } yield {
      val definitions = new scala.collection.mutable.ArrayBuffer[ClassWithPos]()

      def name(tree: Tree) = tree match {
        case cls: Defn.Class => s"Class ${cls.name.value}"
        case cls: Defn.Object => s"Object ${cls.name.value}"
        case cls: Defn.Trait => s"Trait ${cls.name.value}"
        case cls: Defn.Enum => s"Enum ${cls.name.value}"
      }
      def loop(tree: Tree, isToplevel: Boolean = true): Unit = tree match {
        case _: Pkg | _: Pkg.Object =>
          tree.children.foreach(loop(_))
        case _: Defn.Class | _: Defn.Trait | _: Defn.Object | _: Defn.Enum =>
          definitions.append(ClassWithPos(name(tree), tree.pos))
          if (checkInnerClasses)
            tree.children.foreach(loop(_, isToplevel = false))
        case dfn: Defn.Def if isToplevel =>
          definitions.append(ClassWithPos(s"Toplevel package", dfn.pos))
          tree.children.foreach(loop(_, isToplevel = false))
        case _ =>
          tree.children.foreach(loop(_))
      }
      definitions.toList
    }
  }
```

This avoids the additional var and excessive checking in definitions.

Also, you forgot about enum.